lightning/src/util/sweep.rs

```diff
@@ -463,39 +449,91 @@ where
 		self.sweeper_state.lock().unwrap().best_block
 	}
 
-	fn regenerate_spend_if_necessary(
-		&self, sweeper_state: &mut SweeperState,
-	) -> Option<Transaction> {
-		let cur_height = sweeper_state.best_block.height;
-		let cur_hash = sweeper_state.best_block.block_hash;
-		let filter_fn = |o: &TrackedSpendableOutput| {
-			if o.status.is_confirmed() {
-				// Don't rebroadcast confirmed txs.
-				return false;
-			}
+	/// Regenerates and broadcasts the spending transaction for any outputs that are pending
+	pub async fn regenerate_and_broadcast_spend_if_necessary_locked(&self) -> Result<(), ()> {
+		// Collect spendable output descriptors.
+		let respend_descriptors_clones: Vec<SpendableOutputDescriptor>;
+		let respend_descriptors: Vec<&SpendableOutputDescriptor>;
+		{
+			let mut sweeper_state = self.sweeper_state.lock().unwrap();
 
-			if o.status.is_delayed(cur_height) {
-				// Don't generate and broadcast if still delayed
-				return false;
+			// Prevent concurrent sweeping.
+			if sweeper_state.sweep_pending {
```

Comment:
Added this boolean. Seemed easiest with a locked state already.

Reply:
Alternative option with AtomicBool might create a nested locks situation that can be risky?

Reply:
To me, state isn't necessarily limited to persistent state. I could live with some items persisted and others not. The AtomicBool option is locking unnecessarily perhaps, with nesting potentially too. Maybe I could split out the persistent part of the state in a separate struct, and group that with the runtime state. Then still a single mutex can guard all of it?

Reply:
Converted to using AtomicBool now.